quickwit-indexing/src/source/kafka_source.rs

```diff
@@ -446,6 +476,49 @@ impl Source for KafkaSource {
     }
 }
 
+#[derive(Debug)]
+pub(crate) enum KafkaMessage {
+    Message {
+        doc_opt: Option<String>,
+        payload_len: u64,
+        partition: i32,
+        offset: i64,
+        _timestamp: Timestamp,
+    },
+    PartitionEOF(i32),
+    Err(anyhow::Error),
+}
+
+impl From<KafkaResult<BorrowedMessage<'_>>> for KafkaMessage {
+    fn from(message_res: KafkaResult<BorrowedMessage<'_>>) -> Self {
+        match message_res {
+            Ok(message) => Self::Message {
+                doc_opt: parse_message_payload(&message),
+                payload_len: message.payload_len() as u64,
+                partition: message.partition(),
+                offset: message.offset() as i64,
+                _timestamp: message.timestamp(),
+            },
+            Err(KafkaError::PartitionEOF(partition)) => Self::PartitionEOF(partition),
+            Err(error) => Self::Err(anyhow::anyhow!(error)),
+        }
+    }
+}
+
+fn spawn_consumer_poll_loop(
+    consumer: Arc<RdKafkaConsumer>,
+) -> (JoinHandle<()>, flume::Receiver<KafkaMessage>) {
+    let (messages_tx, messages_rx) = flume::bounded(2);
+    let join_handle = spawn_blocking(move || {
+        while let Some(message) = consumer.poll(None) {
```

Comment:
So in trying to test this I'm now seeing the code block on this line and not process data, "most of the time". I have seen it process some data a couple times but I'm not sure what the magic incantation was for that to work. 

This indirection and message passing between threads seems kinda complex and problematic to me. I believe it is the result from trying to remove the `futures::executor::block_on` calls but all those were doing is blocking the rebalance thread until the async operations complete which is the behavior you want in that instance anyway. Obviously I'm not an expert on Rust or tokio so maybe there's some side effect in the simpler implementation that I'm not aware of, but looking at how this is behaving now it's hard to see what advantage it has and it seems to be kinda falling down a rabbit hole of complexity.

Reply:
I have the same issue. The code does not block. The consumer is simply waiting for new messages... Indeed, moving the reassignment logic from the pre-rebalance hook to the post-rebalance one does not work. Mutating the rebalance object in this listener is no-op; see this [code](https://github.com/fede1024/rust-rdkafka/blob/master/src/consumer/mod.rs#L96). With the current code, the consumer resumes at the offset dictated by the reset policy (`auto.offset.reset`) after the rebalance. The default is `latest`.

For our use case, the Kafka Definitive Guide recommends seeking after the rebalance. So I tried that, and I hit two bugs in librdkafka. The current recommendation of the maintainers librdkafka is to simply not seek with consumer groups... The Java client seems to work fine, though. So I will open an issue and make a case for our cause. Otherwise, we will have to go with another strategy in the pre-rebalance where we lock each partition before consumption and receive a "publish token", which you already suggested in the past.

Regarding consuming messages in a tokio block task, I'm not trying to avoid `futures::executor::block_on`. Actually, the new calls to `send` and `recv` that I introduced in the pre and post-rebalance hooks are also blocking.

I'm trying to achieve two things:
1. Avoid blocking the tokio main loop
While `consumer.recv()` returns a future and seems non-blocking, this is not always the case. Sometimes, when calling `consumer.recv()`, the Kafka group coordinator informs the consumer that a rebalance is required, and the consumer calls the rebalance code right away in the same future. Unfortunately, the rebalance code is sync and while it is executing, the tokio main loop does not make any progress; hence the lock I observed in my first attempt. There are ways to write some "smart blocking code" that does not block for too long, but that's a risky endeavor.

2. Avoid sharing the state of the consumer with the consumer context
On pre or post-rebalance, we will eventually want to update the state of the consumer (increment the indexing generation, clear the current in-flight batch of docs, update some stats.), which means we will have to share the state of the consumer with the context via an `Arc<Mutex<>>` or `Arc<RwLock<>>`. Dealing with locks is hard. I find the message passing way much cleaner: first, the context sends a message to the consumer and waits for an ack; second, the consumer handles the message with unrestricted access to its state thanks to its `&mut self` methods, and finally acks. Some additional benefits are:
- it's easy to reason about the order of the events
- the rebalance logic exists in one place instead of being spread across two places
- the rebalance logic is easier to test

tl;dr 
Bear with me while I'm trying to make the seek offsets after rebalancing strategy work.

Reply:
Also, I measured the message passing overhead on my machine:
- `main` branch: 250-400k messages/s
- this branch: 100k messages/s (channel size: 2)
- tweaking the size of the channel between the consumer and the actor: up to 1.2M messages/s with a channel size of 1,000.